service/controller/v3/resource/chart/desired.go

```diff
@@ -65,3 +73,43 @@ func (r *Resource) getConfigMapValues(ctx context.Context, customObject v1alpha1
 
 	return chartValues, nil
 }
+
+func (r *Resource) getSecretValues(ctx context.Context, customObject v1alpha1.ChartConfig) (map[string]interface{}, error) {
+	secretValues := make(map[string]interface{})
+
+	if key.SecretName(customObject) != "" {
+		secretName := key.SecretName(customObject)
+		secretNamespace := key.SecretNamespace(customObject)
+
+		secret, err := r.k8sClient.CoreV1().Secrets(secretNamespace).Get(secretName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return make(map[string]interface{}), microerror.Maskf(notFoundError, "secret '%s' in namespace '%s' not found", secretName, secretNamespace)
+		} else if err != nil {
+			return make(map[string]interface{}), microerror.Mask(err)
+		}
+
+		// TODO: fix this "secret.json" name somewhere and access it in release-operator.
+		secretData := secret.Data["secret.json"]
+		if secretData != nil {
+			err = json.Unmarshal(secretData, &secretValues)
+			if err != nil {
+				return secretValues, microerror.Mask(err)
+			}
+		}
+	}
+
+	return secretValues, nil
+}
+
+func union(a, b map[string]interface{}) (map[string]interface{}, error) {
```

Comment:
This function does a simple union of two maps. The resulting map should have all key/value pairs of the inputs.
We error if the key is shared between inputs, our configmap and secret is supposed to have mutually exclusive values.

Reply:
This is cool and needed but I think it should have separate tests. As if it goes wrong someone is going to have a bad day debugging it!

Reply:
+1 we need unit tests for this.

Reply:
I wasn't thinking here. Added unit tests and discovered two fixes for `nil` input along the way.

Thanks for keeping an eye out for my foolery :+1: